spec.bs

```diff
@@ -0,0 +1,369 @@
+    1. If |topic| is not |undefined|, add |topic| to |result|.
+  1. Remove duplicate entries in |result|.
+  1. Random shuffle entries in |result|.
+  1. Return |result|.
+
+</section>
+
+<section>
+  <h2 id="the-javascript-api-header">The JavaScript API</h2>
+
+  The Topics API lives under the {{Document}} interface, and is only available if the document is in [=secure context=].
+
+  <pre class="idl">
+    dictionary BrowsingTopicsOptions {
+      boolean observe = true;
+    };
+
+    partial interface Document {
+        Promise&lt;sequence&lt;BrowsingTopic&gt;&gt; browsingTopics(optional BrowsingTopicsOptions options);
+    };
+  </pre>
+
+
+  The <dfn for="Document" method>browsingTopics(options)</dfn> method steps are:
+  1. Let |p| be [=a new promise=].
+  1. Run the following steps [=in parallel=]:
+    1. Determine the [=topics-caller-context=] |C|.
+    1. If any of the following is true:
+        - [=this=] is not [=allowed to use=] the <code><a href="#browsing-topics-policy-controlled-feature">browsing-topics</a></code> feature.
+        - [=this=] is not [=allowed to use=] the <code><a href="https://wicg.github.io/floc/#interest-cohort-policy-controlled-feature">interest-cohort</a></code> feature.
+        - The user preference setting disallows the access to topics from |C|.caller-origin and |C|.page.top-level-origin.
+
+        then:
+          1. [=Queue a global task=] on the <dfn>browsing topics task source</dfn> given [=this=]'s [=relevant global object=] to [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
+          1. Abort these steps.
+    1. Let |result| be the result from running the [=topics for caller=] algorithm, with |C| as input.
+    1. If options.observe is true:
+        1. Run the [=collect page topics calculation input data=] algorithm with |C|.page as input.
+        1. Run the [=collect topics caller origin=] algorithm with |C|.page and |C|.caller-origin as input.
+    1. [=Queue a global task=] on the [=browsing topics task source=] given [=this=]'s [=relevant global object=] to perform the following steps:
+        1. [=Resolve=] |p| with |result|.
+  1. Return |p|.
+
+</section>
+
+<section>
+  <h2 id="http-request-header-header">HTTP Request Header</h2>
+
+  Topics can be send in the HTTP header for navigation and fetch() request. The corresponding response header is able to specify whether the caller should to be recorded.
+
+  <h3 id="topics-allowable-http-request">Topics-allowable HTTP request</h3>
+
+  <h4 id="topics-allowable-navigation-request-header">Topics-allowable navigation request</h4>
+  The following algorithm determines whether topics are allowed to be sent on a navigation request header:
+    1. Determine the [=topics-caller-context=] |C|.
+    1. If |C|.caller-origin is not <a href="https://www.w3.org/TR/secure-contexts/#potentially-trustworthy-origin">potentially trustworthy</a>:
+        1. Return false.
+    1. If the navigation happens on an iframe:
```

Comment:
This is not precisely defined. It would likely be easier if you had a proper patch to HTML, and then your navigation took appropriate inputs to determine that.